### PR TITLE
Do not call component update on failed components

### LINKED ISSUE #1628: ssd1306_i2c Bootloop
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Hass.io
**ESP (ESP32/ESP8266, Board/Sonoff):**

NodeMCU ESP8266
**ESPHome version (latest production, beta, dev branch)**

v1.15.3 and v1.16.0-dev
**Affected component:**

[ssd1306_i2c](https://esphome.io/components/display/ssd1306.html)

**Description of problem:**
When the display is not detected, the ESP boot loops. I was able to reproduce this config with a minimal config shown below. It appears to be related to the interval attempting to change the display, but it does not exist since the display was not able to be configured. When the interval is removed, the ESP boots up fine. I believe that there needs to be a check in the interval to ensure that the display exists and is set up before proceeding. 

**Problem-relevant YAML-configuration entries:**
```yaml
esphome:
  name: test
  platform: ESP8266
  board: esp12e

# Enable logging
logger:
    
i2c:
  sda: 4
  scl: 5
  scan: True

font:
  - file: "arial.ttf"
    id: arial
    size: 12

display:
  - platform: ssd1306_i2c
    model: "SSD1306 128x32"
    address: 0x3C
    id: display_1
    pages:
      - id: page1
        lambda: |-
          it.print(0, 0, id(arial), "test1");
      
      - id: page2
        lambda: |-
          it.print(0, 0, id(arial), "test2");

interval:
  - interval: 5s
    then:
      - display.page.show_next: display_1
      - component.update: display_1
```

**Logs (if applicable):**

```
[21:29:36]
[21:29:36] ets Jan  8 2013,rst cause:2, boot mode:(3,6)
[21:29:36]
[21:29:36]load 0x4010f000, len 3584, room 16 
[21:29:36]tail 0
[21:29:36]chksum 0xb0
[21:29:36]csum 0xb0
[21:29:36]v2843a5ac
[21:29:36]~ld
[21:29:36][I][logger:166]: Log initialized
[21:29:36][I][app:029]: Running through setup()...
[21:29:36][C][ssd1306_i2c:010]: Setting up I2C SSD1306...
[21:29:36][W][i2c:070]: Received NACK on transmit of address 0x3C
[21:29:36][E][component:092]: Component was marked as failed.
[21:29:36][I][app:059]: setup() finished successfully!
[21:29:36]
[21:29:36]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[21:29:36]
[21:29:36]Exception (29):
[21:29:36]epc1=0x40201ee2 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000
[21:29:36]
[21:29:36]>>>stack>>>
[21:29:36]
[21:29:36]ctx: cont
[21:29:36]sp: 3ffffcc0 end: 3fffffc0 offset: 0190
[21:29:36]3ffffe50:  00000000 3fff032c 00000000 40201eed  
[21:29:36]3ffffe60:  3fff044c 3fff04dc 3fff032c 40201fe8  
[21:29:36]3ffffe70:  00000007 00000009 3ffe8a00 40201029  
[21:29:36]3ffffe80:  00000000 00000004 00000005 402014a8  
[21:29:36]3ffffe90:  3fff033c 4023b49f 00000000 00000004  
[21:29:36]3ffffea0:  00000006 00000007 3ffe87fd 40201cb9  
[21:29:36]3ffffeb0:  00000000 4bc6a7f0 3fff041c 40208b46  
[21:29:36]3ffffec0:  00000000 00000000 4bc6a7f0 40208bc4  
[21:29:36]3ffffed0:  00000007 3fff047c 3fff03dc 40208bd8  
[21:29:36]3ffffee0:  00000000 00000004 3ffee660 40208bf8  
[21:29:36]3ffffef0:  3fffdad0 00000000 3ffee660 40208a84  
[21:29:36]3fffff00:  3fffdad0 00000000 3ffee660 401002bc  
[21:29:36]3fffff10:  00000000 4bc6a7f0 0000006d 3ffee9f4  
[21:29:36]3fffff20:  00000000 00000000 4bc6a7f0 00000000  
[21:29:36]3fffff30:  0000006d 3fff04e4 401007ce 69374bc6  
[21:29:36]3fffff40:  00000000 00000004 00000006 3ffee9f4  
[21:29:36]3fffff50:  3fffdad0 00000000 3ffee660 40202501  
[21:29:36]3fffff60:  00000005 00000009 3fff03b4 3fff03f4  
[21:29:36]3fffff70:  00000000 feefeffe feefeffe feefeffe  
[21:29:36]3fffff80:  3fff033c 3ffee6c8 00000708 3ffee9f4  
[21:29:36]3fffff90:  3fffdad0 00000000 3ffee9b4 40204fe0  
[21:29:36]3fffffa0:  3fffdad0 00000000 3ffee9b4 402068ec  
[21:29:36]3fffffb0:  feefeffe feefeffe 3ffe8514 401003b9  
[21:29:36]<<<stack<<<
[21:29:36]
[21:29:36]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[21:29:36]
[21:29:36] ets Jan  8 2013,rst cause:2, boot mode:(3,6)
[21:29:36]
[21:29:36]load 0x4010f000, len 3584, room 16 
[21:29:36]tail 0
[21:29:36]chksum 0xb0
[21:29:36]csum 0xb0
[21:29:36]v2843a5ac
[21:29:36]~ld
[21:29:36][I][logger:166]: Log initialized
[21:29:36][I][app:029]: Running through setup()...
[21:29:36][C][ssd1306_i2c:010]: Setting up I2C SSD1306...
[21:29:36][W][i2c:070]: Received NACK on transmit of address 0x3C
[21:29:36][E][component:092]: Component was marked as failed.
[21:29:36][I][app:059]: setup() finished successfully!
[21:29:36]
[21:29:36]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[21:29:36]
[21:29:36]Exception (29):
[21:29:36]epc1=0x40201ee2 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000
[21:29:36]
[21:29:36]>>>stack>>>
[21:29:36]
[21:29:36]ctx: cont
[21:29:36]sp: 3ffffcc0 end: 3fffffc0 offset: 0190
[21:29:36]3ffffe50:  00000000 3fff032c 00000000 40201eed  
[21:29:36]3ffffe60:  3fff044c 3fff04dc 3fff032c 40201fe8  
[21:29:36]3ffffe70:  00000007 00000009 3ffe8a00 40201029  
[21:29:36]3ffffe80:  00000000 00000004 00000005 402014a8  
[21:29:36]3ffffe90:  3fff033c 4023b49f 00000000 00000004  
[21:29:36]3ffffea0:  00000006 00000007 3ffe87fd 40201cb9  
[21:29:36]3ffffeb0:  00000000 4bc6a7f0 3fff041c 40208b46  
[21:29:36]3ffffec0:  00000000 00000000 4bc6a7f0 40208bc4  
[21:29:36]3ffffed0:  00000007 3fff047c 3fff03dc 40208bd8  
[21:29:36]3ffffee0:  00000000 00000004 3ffee660 40208bf8  
[21:29:36]3ffffef0:  3fffdad0 00000000 3ffee660 40208a84  
[21:29:36]3fffff00:  3fffdad0 00000000 3ffee660 401002bc  
[21:29:36]3fffff10:  00000000 4bc6a7f0 0000006c 3ffee9f4  
[21:29:36]3fffff20:  00000000 00000000 4bc6a7f0 00000000  
[21:29:36]3fffff30:  0000006c 3fff04e4 401007ce a083126e  
[21:29:36]3fffff40:  00000000 00000004 00000006 3ffee9f4  
[21:29:36]3fffff50:  3fffdad0 00000000 3ffee660 40202501  
[21:29:36]3fffff60:  00000005 00000009 3fff03b4 3fff03f4  
[21:29:36]3fffff70:  00000000 feefeffe feefeffe feefeffe  
[21:29:36]3fffff80:  3fff033c 3ffee6c8 00000708 3ffee9f4  
[21:29:36]3fffff90:  3fffdad0 00000000 3ffee9b4 40204fe0  
[21:29:36]3fffffa0:  3fffdad0 00000000 3ffee9b4 402068ec  
[21:29:36]3fffffb0:  feefeffe feefeffe 3ffe8514 401003b9  
[21:29:36]<<<stack<<<
[21:29:36]
[21:29:36]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[21:29:36]
```

**Additional information and things you've tried:**
I've tried the minimal config on both the latest release and dev with the same issues.

### PR DESCRIPTION
**Pull request by alexyao2015**
## Description:
Avoids updating the display if not attached. This same issue likely affects other display devices such as ssd1325, but this only fixes ssd1306.

**Related issue (if applicable):** fixes esphome/issues#1628

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by kbx81** [comment]
This makes sense and looks good to me. I'll try to make some time to test on my '1306, as well. 😄

**Comment by alexyao2015** [comment]
Perhaps something should be built into the component class instead that calls display so that failed components are checked automatically. This would be similar to the current loop implementation. This likely isn't the best implementation and could be refactored to be more deeply integrated.

**Comment by kbx81** [comment]
Agreed -- this is more of a "quick and dirty" fix. Digging deeper down into the stack would be a better solution but could also impact other I2C components...it will likely result in a lot more work and a bigger PR. It's not a problem, it just depends on how much more work we want to create. 😄 We are pretty backlogged in terms of PRs...

**Review by jesserockz — COMMENTED** [review]
I feel like this is something that should be done at the component level here:
https://github.com/esphome/esphome/blob/b52f7cfe86a167bcd3f41baf2194243699a3e354/esphome/core/component.cpp#L163
so if a component `is_failed()` it should not call `update`. It already skips calling `loop` if its failed.

**Comment by alexyao2015** [comment]
Agreed! I'll look into changing this.

File: esphome/components/ssd1306_base/ssd1306_base.cpp [hunk 1]
```diff
@@ -125,8 +125,10 @@ bool SSD1306::is_sh1106_() const {
          this->model_ == SH1106_MODEL_128_64;
 }
 void SSD1306::update() {
-  this->do_update_();
-  this->display();
+  if (!this->is_failed()) {
```
Comment (ssieb) [inline esphome/components/ssd1306_base/ssd1306_base.cpp:128]: I would suggest a cleaner way would be to instead use:
```
if (this->is_failed())
  return;
```

**Comment by alexyao2015** [comment]
I've now updated this to fix all display bootloop. A lingering issue is below OTA crashing during safe mode but not user mode only when a font is added and is dependent on the number of glyphs. If I only add 1 glyph and forcibly enter safe mode through the rst button, it ota's fine. But with the default glyphs or even reducing the glyphs, it still bootloops in safe mode, but not in user mode. 
Here is a decoded stack trace along with the yaml if anyone here is able to take a look. https://hatebin.com/pmnrnlwequ

File: esphome/components/ssd1306_base/ssd1306_base.cpp (hunk 1 above)
Comment (alexyao2015) [reply esphome/components/ssd1306_base/ssd1306_base.cpp]: Did that with the new revision!

**Comment by alexyao2015** [comment]
The lingering issue was fixed in #1534

**Comment by glmnet** [comment]
So is this actually about calling `update()` on failed components?

Why isn't [this](https://github.com/esphome/esphome/pull/1392#pullrequestreview-589959119) also fixed?

**Comment by alexyao2015** [comment]
Everything ends up going through that automation, but not everything goes through component is what I learned through testing. Interval specifically directly calls the automation which crashes it.

**Comment by glmnet** [comment]
No, if your component inherits `PollingComponent` the interval is setup there and has nothing to with the UpdateComponentAction

Otherwise can you explain exactly whats going on?

**Comment by alexyao2015** [comment]
Here is a test yaml and short snippet of the main.cpp. You can see that interval directly calls `UpdateComponentAction` https://hatebin.com/bazyfpebbr

**Comment by glmnet** [comment]
The interval component is fine, you don't need to mark it failed.

You're wrong about the `update()` call being routed to the `UpdateComponentAction`. See, the `set_interval` Jesse points to is a ESPHome scheduler (thus magic going on here again), this is not calling update because of this:

https://github.com/esphome/esphome/blob/46356cbc4afbb2e38380e4702919927d40ea2192/esphome/core/scheduler.cpp#L143-L147

So this PR is ok on fixing the `UpdateComponentAction` (still you could call `.update()` on a lambda and crash it though)

I'm not so sure about the priority change, 399 is an awful number and this can be a behavior change for weird use case?

So I'd recommend removing that part (you can create another PR with it's own discussion) and rename this to what it really does (e.b. Do not call component update on failed components) then this can go as a bug/fix asap)

File: esphome/core_config.py [hunk 2]
```diff
@@ -160,7 +160,7 @@ def valid_include(value):
         cv.Optional(CONF_ON_BOOT): automation.validate_automation(
             {
                 cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(StartupTrigger),
-                cv.Optional(CONF_PRIORITY, default=600.0): cv.float_,
+                cv.Optional(CONF_PRIORITY, default=399.0): cv.float_,
```
Comment (jesserockz) [inline esphome/core_config.py:163]: Reverting this so we can put the update fix in
```suggestion
                cv.Optional(CONF_PRIORITY, default=600.0): cv.float_,
```

**Comment by alexyao2015** [comment]
@jesserockz Something still should be changed about the priority again. See https://esphome.io/components/esphome.html?highlight=on_boot#esphome-on-boot

Also the display.update crashes in on_boot if it is at 600 since that has a priority of 400 I think.

**Comment by jesserockz** [comment]
Yes, but lets merge this and we can discuss that change in another PR/discussion again =)

**Review by jesserockz — APPROVED** [review]

**Pull request merged**